tests/unit/s2n_protocol_preferences_test.c

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+#include "tls/s2n_connection.h"
+
+#include <s2n.h>
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Test config append */
+    {
+        struct s2n_config *config;
+        EXPECT_NOT_NULL(config = s2n_config_new());
+        EXPECT_EQUAL(config->application_protocols.size, 0);
+        size_t prev_size = 0;
+
+        /* should grow the blob with the provided value */
+        EXPECT_SUCCESS(s2n_config_append_protocol_preference(config, (const uint8_t *)"protocol1", 9));
+        EXPECT_EQUAL(config->application_protocols.size, 1 /* len prefix */ + 9);
+        prev_size = config->application_protocols.size;
+
+        /* should grow the blob even more */
+        EXPECT_SUCCESS(s2n_config_append_protocol_preference(config, (const uint8_t *)"protocol2", 9));
+        EXPECT_EQUAL(config->application_protocols.size, prev_size + 1 /* len prefix */ + 9);
+        prev_size = config->application_protocols.size;
+
+        /* should allow null byte values */
+        const uint8_t null_value[9] = { 0 };
+        EXPECT_SUCCESS(s2n_config_append_protocol_preference(config, (const uint8_t *)null_value, 9));
+        EXPECT_EQUAL(config->application_protocols.size, prev_size + 1 /* len prefix */ + 9);
+        prev_size = config->application_protocols.size;
+
+        /* should reallocate the blob */
+        const uint8_t large_value[255] = { 0 };
+        EXPECT_SUCCESS(s2n_config_append_protocol_preference(config, (const uint8_t *)large_value, 255));
+        EXPECT_EQUAL(config->application_protocols.size, prev_size + 1 /* len prefix */ + 255);
```

Comment:
Use sizeof(large_value) instead of 255. You also don't need to cast here.

Like I'd write each of these as:
```
        const uint8_t value[] = "value"; // or const uint8_t value[size] = { 0 };
        EXPECT_SUCCESS(s2n_config_append_protocol_preference(config, value, sizeof(value));
        EXPECT_EQUAL(config->application_protocols.size, prev_size + LEN_PREFIX + sizeof(value);
```

Reply:
> You also don't need to cast here.

I do since arrays are different types:

```
/home/s2n-dev/s2n/api/s2n.h:233:92: note: expected 'const uint8_t *' {aka 'const unsigned char *'} but argument is of type 'const uint8_t (*)[255]' {aka 'const unsigned char (*)[255]'}
  233 | nfig_append_protocol_preference(struct s2n_config *config, const uint8_t *protocol, uint8_t protocol_len);
      |                                                            ~~~~~~~~~~~~~~~^~~~~~~~
```

I'll use the size_of, though

Reply:
Scratch that; I was passing `&large_value` and not `large_value`. I'll fix it.